crates/crypto/keystore/src/encrypted_keystore.rs

```diff
@@ -0,0 +1,206 @@
+use std::{fs, path::Path};
+
+use alloy_primitives::hex::{FromHex, ToHexExt};
+use anyhow::Result;
+use ream_bls::PubKey;
+use serde::{Deserialize, Deserializer, Serialize, Serializer, de};
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct EncryptedKeystore {
+    pub crypto: Crypto,
+    pub description: String,
+    pub pubkey: PubKey,
+    pub path: String,
+    pub uuid: String,
+    pub version: u64,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct Crypto {
+    pub kdf: FunctionBlock<KdfParams>,
+    pub checksum: FunctionBlock<ChecksumParams>,
+    pub cipher: FunctionBlock<CipherParams>,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct FunctionBlock<ParamType> {
+    #[serde(flatten)]
+    pub params: ParamType,
+    #[serde(serialize_with = "buffer_to_hex", deserialize_with = "hex_to_buffer")]
+    pub message: Vec<u8>,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+#[serde(tag = "function", content = "params")]
+pub enum KdfParams {
+    #[serde(rename = "pbkdf2")]
+    Pbkdf2 {
+        c: u32,
+        dklen: u8,
+        prf: String,
+        #[serde(serialize_with = "buffer_to_hex", deserialize_with = "hex_to_buffer")]
+        salt: Vec<u8>,
+    },
+    #[serde(rename = "scrypt")]
+    Scrypt {
+        dklen: u8,
+        n: u32,
+        p: u32,
+        r: u32,
+        #[serde(serialize_with = "buffer_to_hex", deserialize_with = "hex_to_buffer")]
+        salt: Vec<u8>,
+    },
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+#[serde(tag = "function", content = "params")]
+pub enum CipherParams {
+    #[serde(rename = "aes-128-ctr")]
+    Aes128Ctr {
+        #[serde(serialize_with = "buffer_to_hex", deserialize_with = "hex_to_buffer")]
+        iv: Vec<u8>,
+    },
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+#[serde(tag = "function", content = "params")]
+pub enum ChecksumParams {
+    #[serde(rename = "sha256")]
+    Sha256 {},
+}
+
+fn buffer_to_hex<T, S>(buffer: &T, serializer: S) -> Result<S::Ok, S::Error>
+where
+    T: AsRef<[u8]>,
+    S: Serializer,
+{
+    serializer.serialize_str(&buffer.encode_hex())
+}
+
+fn hex_to_buffer<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    String::deserialize(deserializer).and_then(|string| {
+        <Vec<u8>>::from_hex(&string).map_err(|err| de::Error::custom(err.to_string()))
+    })
+}
+
+impl EncryptedKeystore {
+    pub fn load_from_file<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let file_contents = fs::read_to_string(path)?;
+        let keystore: EncryptedKeystore = serde_json::from_str(&file_contents)?;
+
+        Ok(keystore)
+    }
+
+    pub fn save_to_file<P: AsRef<Path>>(&self, path: P) -> Result<()> {
+        let serialized = serde_json::to_string(self)?;
```

Comment:
the struct and impl definition should be right next to each other, not separate by random objects

```rust
enum X {}

impl X;

```

Don't put objects between these, ideally put the other objects below the object and impl definitions